.github/workflows/build_and_publish.yml

```diff
@@ -0,0 +1,137 @@
+name: Publish tagged releases to PyPI
+
+on:
+  release:
+    types:
+      - created
+
+jobs:
+  run_tests:
+    runs-on: ubuntu-latest
+
+    strategy:
+      fail-fast: true
+      matrix:
+        python-version: ["3.9", "3.10", "3.11"]
+
+    steps:
+      - uses: actions/checkout@v3
+
+      - name: Setup Python
+        uses: actions/setup-python@v3
+        with:
+          python-version: ${{ matrix.python-version }}
+
+      - name: Install poetry
+        uses: snok/install-poetry@v1
+        with:
+          version: latest
+          virtualenvs-create: true
+          virtualenvs-in-project: true
+          installer-parallel: true
+
+      - name: Load cache
+        uses: actions/cache@v3
+        id: cache
+        with:
+          path: .venv
+          key: venv-${{ runner.os }}-${{ matrix.python-version }}-${{ hashFiles('**/pyproject.toml') }}
+
+      - name: Install dependencies
+        if: steps.cache.outputs.cache-hit != 'true'
+        run: poetry install --no-interaction
+
+      - name: Run mypy
+        run: poetry run mypy src
+
+      - name: Check styling
+        run: poetry run black . --check
+
+      - name: Linting
+        run: poetry run ruff check .
+
+      - name: Run tests
+        run: poetry run pytest
+
+  publish_to_testpypi:
+    name: Build source distribution and publish to test-PyPI
+    needs: [run_tests]
+    runs-on: ubuntu-latest
+
+    strategy:
+      fail-fast: true
+
+    steps:
+      - uses: actions/checkout@v3
+
+      - name: Setup Python
+        uses: actions/setup-python@v3
+        with:
+          python-version: 3.9
+
+      - name: Install poetry
+        uses: snok/install-poetry@v1
+        with:
+          version: latest
+          virtualenvs-create: true
+          virtualenvs-in-project: true
+          installer-parallel: true
+
+      - name: Load cache
+        uses: actions/cache@v3
+        id: cache
+        with:
+          path: .venv
+          key: venv-${{ runner.os }}-3.9-${{ hashFiles('**/pyproject.toml') }}
+
+      - name: Install dependencies
+        if: steps.cache.outputs.cache-hit != 'true'
+        run: poetry install --no-interaction
+
+      - name: Setup test-PyPI credentials
+        run: |
+          poetry config repositories.testpypi https://test.pypi.org/legacy/
+          poetry config pypi-token.testpypi ${{ secrets.TEST_PYPI_TOKEN }}
+
+      - name: Publish to test-PyPI
+        run: poetry publish -r testpypi
+
+  publish_to_pypi:
```

Comment:
should be publish_to_test_pypi, since the 'scripts' all seem to go to test-pypi ?

Reply:
It's actually PyPI. The workflow goes test -> publish to test-pypi -> publish to pypi. I did misspecify the name. Will correct that lol